source/common/http/alternate_protocols.cc

```diff
@@ -0,0 +1,40 @@
+#include "common/http/alternate_protocols.h"
+
+namespace Envoy {
+namespace Http {
+
+AlternateProtocols::AlternateProtocol::AlternateProtocol(absl::string_view alpn,
+                                                         absl::string_view hostname, int port)
+    : alpn_(alpn), hostname_(hostname), port_(port) {}
+
+AlternateProtocols::Origin::Origin(absl::string_view scheme, absl::string_view hostname, int port)
+    : scheme_(scheme), hostname_(hostname), port_(port) {}
+
+AlternateProtocols::AlternateProtocols(TimeSource& time_source) : time_source_(time_source) {}
+
+void AlternateProtocols::setAlternatives(const Origin& origin,
+                                         const std::vector<AlternateProtocol>& protocols,
+                                         const MonotonicTime& ttl) {
+  protocols_[origin] = {protocols, ttl};
```

Comment:
probably premature optimization but won't we basically then update the entire vector on every response header? Would it make more sense to just update the ttl if it's new?